validator/src/journal/block_manager.rs

```diff
@@ -269,16 +299,41 @@ impl BlockManagerState {
     }
 
     fn ref_block(&self, block_id: &str) -> Result<(), BlockManagerError> {
-        match self
+        let mut references_by_block_id = self
             .references_by_block_id
             .write()
-            .expect("Acquiring references write lock; lock poisoned")
-            .get_mut(block_id)
-        {
-            Some(r) => r.increase_external_ref_count(),
-            None => return Err(BlockManagerError::UnknownBlock),
+            .expect("Acquiring references write lock; lock poisoned");
+
+        if let Some(r) = references_by_block_id.get_mut(block_id) {
+            r.increase_external_ref_count();
+            return Ok(());
         }
-        Ok(())
+
+        let blockstore_by_name = self
+            .blockstore_by_name
+            .read()
+            .expect("Acquiring blockstore by name read lock; lock poisoned");
+
+        let mut block = None;
+
+        for (_, store) in blockstore_by_name.iter() {
```

Comment:
I would like to see this code-block as finding the store (if it exists) that contains the block. So `blockstore_by_name.iter().find(|(_, store)| store.get(&[block_id]).next().is_some())` or something like that. And then you can do direct expression assignment to block instead of having `mut block`, iteration, and break.

Reply:
That requires getting the block from the store twice, which is why I chose not to do it.

Reply:
`filter_map(|(_, store)| store.get(&[block_id]).next()).next()` may be what is needed.

Reply:
Nice, I think that works